x/stakeibc/keeper/unbonding_records_get_host_zone_unbondings_msgs_test.go

```diff
@@ -1,226 +1,813 @@
-	s.Require().NoError(err, "could unmarshal undelegation callback args")
-
-	// The zero weight validator should not be included in the message
-	expectedNumMessages := len(tc.hostZone.Validators) - 1
-	s.Require().Equal(expectedNumMessages, len(actualCallbackResult.SplitDelegations), "number of split delegations in success unbonding case")
-	s.Require().Equal(tc.hostZone.ChainId, actualCallbackResult.HostZoneId, "host zone id in success unbonding case")
-
-	// TODO add case that checks the *marshaled* callback args against expectations
-
-	// the number of unbonding messages should be (number of validators) * (records to unbond)
-	s.Require().Equal(expectedNumMessages, len(actualUnbondMsgs), "number of unbonding messages should be number of records to unbond")
-
-	s.Require().Equal(tc.amtToUnbond.Mul(sdkmath.NewInt(int64(len(tc.epochUnbondingRecords)))), actualAmtToUnbond, "total amount to unbond should match input amtToUnbond")
+func (s *KeeperTestSuite) TestGetValidatorUnbondCapacity() {
+	// Start with the expected returned list of validator capacities
+	expectedUnbondCapacity := []keeper.ValidatorUnbondCapacity{
+		{
+			ValidatorAddress:   "valA",
+			CurrentDelegation:  sdkmath.NewInt(50),
+			BalancedDelegation: sdkmath.NewInt(0),
+			Capacity:           sdkmath.NewInt(50),
+		},
+		{
+			ValidatorAddress:   "valB",
+			CurrentDelegation:  sdkmath.NewInt(200),
+			BalancedDelegation: sdkmath.NewInt(5),
+			Capacity:           sdkmath.NewInt(195),
+		},
+		{
+			ValidatorAddress:   "valC",
+			CurrentDelegation:  sdkmath.NewInt(1089),
+			BalancedDelegation: sdkmath.NewInt(1000),
+			Capacity:           sdkmath.NewInt(89),
+		},
+	}
 
-	totalWgt := sdk.NewDec(int64(tc.totalWgt))
-	actualAmtToUnbondDec := sdk.NewDecFromInt(actualAmtToUnbond)
-	actualUnbondMsg1 := actualUnbondMsgs[0].String()
-	actualUnbondMsg2 := actualUnbondMsgs[1].String()
-	val1Fraction := sdk.NewDec(int64(tc.hostZone.Validators[0].Weight)).Quo(totalWgt)
-	val2Fraction := sdk.NewDec(int64(tc.hostZone.Validators[1].Weight)).Quo(totalWgt)
-	val1UnbondAmt := val1Fraction.Mul(actualAmtToUnbondDec).TruncateInt().String()
-	val2UnbondAmt := val2Fraction.Mul(actualAmtToUnbondDec).TruncateInt().String()
+	// Build list of input validators and map of balanced delegations from expected list
+	validators := []*types.Validator{}
+	balancedDelegations := map[string]sdkmath.Int{}
+	for _, validatorCapacity := range expectedUnbondCapacity {
+		validators = append(validators, &types.Validator{
+			Address:    validatorCapacity.ValidatorAddress,
+			Delegation: validatorCapacity.CurrentDelegation,
+		})
+		balancedDelegations[validatorCapacity.ValidatorAddress] = validatorCapacity.BalancedDelegation
+	}
 
-	val1Unbonded := strings.Contains(actualUnbondMsg1, val1UnbondAmt)
-	val2Unbonded := strings.Contains(actualUnbondMsg2, val2UnbondAmt)
+	// Add validators with no capacity - none of these should be in the returned list
+	deficits := []int64{0, 10, 50}
```

Comment:
This is the first test that is hard to understand... what is deficits?
So you set up 6 validators, the first three have capacity, but the last 3 are off by these positive deficits meaning they can't actually be affected by an unbonding...
But I don't see how the numbers add up... When you add the last 3 validators that would completely change the actual capacities wouldn't it?  Are we assuming the weights would be such that this all works out so we take it as granted the capacities were already computed correctly?

Reply:
This is the test for the subfunction `GetValidatorUnbondCapacity`. So all it checks is the difference between the balanced delegation and the current delegation. It doesn't consider the total stake at all.

any suggestion on how I could make the deficit variable more clear? Or maybe I just extend the comment?

Reply:
Yeah I was confused because the capacity calculations have to depend on the total and the unbond amount and the weights but that isn't what this test is trying to test at all.  You aren't worried about the weights and we can take as given that the capacities are already correct and set.  I was confused about the seemingly arbitrary numbers in deficits but I was missing the point